analysers/analyser_osmosis_highway_deadend.py

```diff
@@ -71,121 +71,227 @@
+"""
+
+sql32 = """
+-- Collect all nodes that could be possible (valid) end nodes
+CREATE TEMP TABLE allowed_end_nodes AS
+SELECT DISTINCT
+  unnest(nodes) AS id
+FROM
+  ways
+WHERE
+  ways.tags != '' AND
+  (
+    (ways.tags?'amenity' AND ways.tags->'amenity' = 'parking') OR
+    (ways.tags?'railway' AND ways.tags->'railway' = 'platform') OR
+    (ways.tags?'aeroway' AND ways.tags->'aeroway' = 'taxiway')
+  )
+UNION ALL
+SELECT
+  id
+FROM
+  nodes
+WHERE
+  tags != '' AND
+  (
+    (tags?'amenity'  AND tags->'amenity' = 'parking_entrance') OR
+    (tags?'entrance' AND tags->'entrance'!= 'garage')
+  )
+UNION ALL
+SELECT DISTINCT
+  unnest(nodes) AS id
+FROM
+  highways
+WHERE
+  -- Non-oneway highways are valid input nodes for oneways
+  (NOT is_oneway AND NOT is_roundabout) OR
+  -- Force motorways as valid input nodes
+  -- Raceways are commonly isolated from the main network
+  -- Escape (emergency stop) ways are supposed to be not used / dead-ended
+  -- Footway can lead to anything (attraction entrances, ...) so exclude them even if oneway
+  highway IN ('motorway', 'raceway', 'escape', 'footway') OR
+  -- Construction roads are usually temporary, the connections are likely access=no or similar
+  is_construction
+UNION ALL
+-- Include nodes outside the extract; the recursive check will otherwise fail on oneways originating outside of the extract. See #1949
+SELECT DISTINCT
+  way_nodes.node_id
+FROM
+  way_nodes
+  LEFT JOIN nodes AS n ON
+    n.id = way_nodes.node_id
+WHERE
+  n IS NULL
+"""
+
+sql33 = """
+CREATE INDEX idx_allowed_nodes_id ON allowed_end_nodes(id);
+"""
+
+sql34 = """
+-- Make table as short as possible: keep only nodes that are actually nodes of oneways
+CREATE TEMP TABLE allowed_oneway_end_nodes AS
+SELECT DISTINCT
+  allowed_end_nodes.id
+FROM
+  allowed_end_nodes
+  JOIN oneway_highway ON
+    -- Note that we cannot literally only get the first and last node, as i.e. in a looped oneway the 'oneway-entrance/exit node' might be in the middle
+    allowed_end_nodes.id = ANY(oneway_highway.nodes)
+"""
+
+sql35 = """
+CREATE INDEX idx_allowed_oneway_nodes_id ON allowed_oneway_end_nodes(id);
+DROP TABLE allowed_end_nodes;
+"""
+
+sql36 = """
+-- All end nodes of oneway streets
+CREATE TEMP TABLE oneway_terminal AS
+SELECT DISTINCT ON (nid)
+  oneway.id AS wid,
+  terminal.id AS nid,
+  terminal.geom AS geom,
+  CASE
+    WHEN oneway_reverse THEN terminal.id = oneway.nodes[1]
+    ELSE terminal.id = oneway.nodes[oneway.length]
+  END AS is_oneway_last_node
+FROM
+  oneway_highway AS oneway
+  JOIN nodes AS terminal ON
+    (
+      terminal.id = oneway.nodes[1] OR
+      terminal.id = oneway.nodes[oneway.length]
+    )
+  LEFT JOIN allowed_oneway_end_nodes ON
+    terminal.id = allowed_oneway_end_nodes.id
+WHERE
+  allowed_oneway_end_nodes IS NULL AND
+  array_length(array_positions(oneway.nodes, terminal.id), 1) = 1 -- exclude end nodes folding back into the same way
```

Comment:
Not sure to understand this condition ? Why not just test oneway is a loop ?

Reply:
Because it can also be a "P-shaped" way. In that case, the point in the middle (where the circle of the P comes back to the vertical line) should be excluded (but not the tip of the "P")

Reply:
Ok, Thank I understand.

Using the start or end node as P junction point depend on the way of the onway. It a detail, it may be ignored.

Reply:
I clarified the comment a bit.

I would personally prefer to keep this in rather than having to use `NOT ST_IsClosed(linestring) AND NOT ST_IsSimple(linestring)`. It seems to be fast, and it's also more accurate than ST_IsSimple about which node to whitelist.